SIL.Windows.Forms.Keyboarding/Linux/GnomeShellIbusKeyboardRetrievingAdaptor.cs

```diff
@@ -10,13 +17,134 @@ namespace SIL.Windows.Forms.Keyboarding.Linux
 	/// The keyboard retrieving part is identical to previous versions but switching keyboards
 	/// changed with 18.04.
 	/// </summary>
-	public class GnomeShellIbusKeyboardRetrievingAdaptor: UnityIbusKeyboardRetrievingAdaptor
+	public class GnomeShellIbusKeyboardRetrievingAdaptor: IbusKeyboardRetrievingAdaptor
 	{
-		protected override IKeyboardSwitchingAdaptor CreateSwitchingAdaptor()
+		private readonly GnomeKeyboardRetrievingHelper _helper = new GnomeKeyboardRetrievingHelper();
+
+		public GnomeShellIbusKeyboardRetrievingAdaptor()
+		{
+		}
+
+		/// <summary>
+		/// Initializes a new instance of the
+		/// <see cref="SIL.Windows.Forms.Keyboarding.Linux.GnomeShellIbusKeyboardRetrievingAdaptor"/> class.
+		/// Used in unit tests
+		/// </summary>
+		/// <param name="ibusCommunicator">Ibus communicator.</param>
+		public GnomeShellIbusKeyboardRetrievingAdaptor(IIbusCommunicator ibusCommunicator): base(ibusCommunicator)
 		{
-			return new GnomeShellIbusKeyboardSwitchingAdaptor(IbusCommunicator);
 		}
 
 		public override bool IsApplicable => _helper.IsApplicable && Platform.IsGnomeShell;
+
+		public override KeyboardAdaptorType Type => KeyboardAdaptorType.System | KeyboardAdaptorType.OtherIm;
+
+		public override void Initialize()
+		{
+			// We come here twice, for both KeyboardType.System and KeyboardType.OtherIm.
+			// We want to create a new switching adaptor only the first time otherwise we're
+			// getting into trouble
+			if (SwitchingAdaptor != null)
+				return;
+
+			SwitchingAdaptor = new GnomeShellIbusKeyboardSwitchingAdaptor(IbusCommunicator);
+			KeyboardRetrievingHelper.AddIbusVersionAsErrorReportProperty();
+			InitKeyboards();
+		}
+
+		protected override void InitKeyboards()
+		{
+			_helper.InitKeyboards(type => true, RegisterKeyboards);
+		}
+
+		private void RegisterKeyboards(IDictionary<string, uint> keyboards, (string type, string layout) firstKeyboard)
```

Comment:
I found I had to keep looking back and forth around this method to try to understand what it was doing, and what the difference is between `keyboards`  and  `curKeyboards`, or which was what meaning. 
I think probably the foreach loop over all ibus keyboards contributed to having trouble understanding at first.

Could we either give a brief method comment saying what the general activity of this method's algorithm is, or sprinkle in some comments inside the method to describe the purpose for sections or some of the contents of the loop. 

It would also help to rename `keyboards` to something like `requestedKeyboards`.

Maybe it could have a method comment that says something like  "Activate all ibus or xkb keyboards in KeyboardController that are requested in the `keyboards` list. Deactivate the rest. Add any keyboards to KeyboardController that were missing."

FWIW, I am leaving my analysis of what the foreach loop looks to be doing:

```
for each ibusKeyboard that ibus knows about: 

  skip processing if it is not in the request list (based on longname or xkb layout name)
  skip processing if we already processed a keyboard with this name/layout.

  if `curKeyboards` contains ibusKeyboard
    make sure it is set as available
    remove it from the `curKeyboards` list so we can later deactivate the rest
  if `curKeyboards` did not contain ibusKeyboard
    add the ibusKeyboard to KeyboardController (but maybe not necessarily to ibus's list of keyboards)
```